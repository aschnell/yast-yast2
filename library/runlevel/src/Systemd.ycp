/**
 * File:	modules/Systemd.ycp
 * Package:	yast2
 * Summary:	systemd configuration
 * Authors:	Ladislav Slez√°k <lslezak@suse.cz>
 *
 * $Id$
 *
 * Functions for setting systemd options
 */

{

module "Systemd";

const string systemd_path = "/bin/systemd";
const string default_symlink = "/etc/systemd/system/default.target";

textdomain "base";

/**
 * Check whether the systemd package is installed
 */
global boolean Installed()
{
    // check for systemd executable
    return SCR::Read(.target.size, systemd_path) >= 0;
}

/*
 * Check whether systemd init is currently running
 * @return boolean true if systemd init is running
 */
global boolean Running()
{
    // check the command of the init program (PID=1)
    const string cmdline_path = "/proc/1/cmdline";

    // /proc/<pid>/cmdline is a NULL separated list
    // we cannot use NULL separator in YCP (?), check just the begining
    string init_cmdline = (string)SCR::Read(.target.string, cmdline_path);
    y2milestone("Current init program: %1", init_cmdline);

    if (init_cmdline == nil)
    {
	y2error("Cannot check systemd status");
	return nil;
    }

    boolean ret = substring(init_cmdline, 0, size(systemd_path)) == systemd_path;
    y2milestone("Systemd detected: %1", ret);

    return ret;
}

/**
 * Set default runlevel for systemd (assumes systemd is installed)
 * @param runlevel the default runlevel to set (integer in range 0..6)
 * @return boolean true on success
 */
global boolean SetDefaultRunlevel(integer runlevel)
{
    if (runlevel == nil || runlevel < 0 || runlevel > 6)
    {
	y2error("Invalid default runlevel (must be in range 0..6): %1", runlevel);
	return false;
    }

    y2milestone("Setting systemd default runlevel: %1", runlevel);

    // create symbolic link, -f to rewrite the current link (if exists)
    string command = sformat("/bin/ln -s -f /lib/systemd/system/runlevel%1.target %2", runlevel, default_symlink);
    y2milestone("Executing: %1", command);

    integer res = (integer)SCR::Execute(.target.bash, command);
    y2debug("Result: %1", res);

    boolean ret = (res == 0);
    y2milestone("Default runlevel set: %1", ret);

    return ret;
}

/**
 * Get the default runlevel for systemd
 * @return integer the default runlevel (or nil on error or unknown runlevel)
 */
global integer DefaultRunlevel()
{
    string target = (string)SCR::Read(.target.symlink, default_symlink);
    y2milestone("Default symlink points to: %1", target);

    if (target == nil)
    {
	y2error("Cannot read symlink target of %1", default_symlink);
	return nil;
    }

    // check runlevel<number>.target
    string runlevel = regexpsub(target, "/runlevel([0-6]).target$", "\\1");
    if (runlevel != nil)
    {
	integer ret = tointeger(runlevel);
	y2milestone("Default runlevel: %1", ret);

	return ret;
    }

    // check runlevel specified by a symbolic name
    // (this is written in systemd documentation how to change the default,
    // YaST should also support this style in case users do a manual change)
    string runlevel_name = regexpsub(target, "/([^/]*).target$", "\\1");
    if (runlevel_name != nil)
    {
	y2milestone("Detected default runlevel name: %1", runlevel_name);
	map<string, integer> mapping = $[
	    "poweroff"	: 0,
	    "rescue"	: 1,
	    // this is ambiguous, runlevels 2 and 4 also point to multi-user
	    // assume runlevel 3 in this case (the most probable)
	    "multi-user": 3,
	    "graphical" : 5,
	    "reboot"	: 6
	];

	integer ret = mapping[runlevel_name]:nil;
	y2milestone("Default runlevel: %1", ret);

	return ret;
    }

    y2error("Cannot determine the default runlevel");
    return nil;
}

} /* EOF */

