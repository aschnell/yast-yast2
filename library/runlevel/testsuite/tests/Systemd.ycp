{
    import "Systemd";
    include "testsuite.ycp";


    DUMP("Systemd::Running() tests");

    // Note: \0 should be used here as the separator, but YCP does not allows that...

    // test standard System V init
    TEST (``(Systemd::Running()), [$[
	    "target" : $[ "string" : "init [5]\n" ]
	    ], $[], $[]], nil);

    // systemd
    TEST (``(Systemd::Running()), [$[
	    "target" : $[ "string" : "/bin/systemd\n" ]
	    ], $[], $[]], nil);

    DUMP("Systemd::SetDefaultRunlevel() tests");

    // test invalid parameters
    TEST (``(Systemd::SetDefaultRunlevel(nil)), [$[], $[], $[
	    "target" : $[ "bash" : 0 ]
	    ]], nil);

    TEST (``(Systemd::SetDefaultRunlevel(-1)), [$[], $[], $[
	    "target" : $[ "bash" : 0 ]
	    ]], nil);

    TEST (``(Systemd::SetDefaultRunlevel(7)), [$[], $[], $[
	    "target" : $[ "bash" : 0 ]
	    ]], nil);

    // test valid parameters
    TEST (``(Systemd::SetDefaultRunlevel(3)), [$[], $[], $[
	    "target" : $[ "bash" : 0 ]
	    ]], nil);

    TEST (``(Systemd::SetDefaultRunlevel(5)), [$[], $[], $[
	    "target" : $[ "bash" : 0 ]
	    ]], nil);

    // test failure
    TEST (``(Systemd::SetDefaultRunlevel(5)), [$[], $[], $[
	    "target" : $[ "bash" : 1 ]
	    ]], nil);


    DUMP("Systemd::DefaultRunlevel() tests");

    // test missing / invalid (not a symlink) default
    TEST (``(Systemd::DefaultRunlevel()), [$[ "target" : $[ "symlink" : nil]], $[], $[]], nil);

    // test numeric runlevel
    TEST (``(Systemd::DefaultRunlevel()), [$[ "target" : $[ "symlink" : "/lib/systemd/system/runlevel3.target"]], $[], $[]], nil);

    // test symbolic runlevel
    TEST (``(Systemd::DefaultRunlevel()), [$[ "target" : $[ "symlink" : "/lib/systemd/system/graphical.target"]], $[], $[]], nil);

    // test unknown symbolic runlevel
    TEST (``(Systemd::DefaultRunlevel()), [$[ "target" : $[ "symlink" : "/lib/systemd/system/unknown_default.target"]], $[], $[]], nil);
}
