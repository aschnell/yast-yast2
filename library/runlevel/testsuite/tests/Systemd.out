Dump	Systemd::Running() tests
Read	.target.string "/proc/1/cmdline" "init [5]\n"
Return	false
Read	.target.string "/proc/1/cmdline" "/bin/systemd\n"
Return	true
Dump	Systemd::SetDefaultRunlevel() tests
Log	Invalid default runlevel (must be in range 0..6): nil
Return	false
Log	Invalid default runlevel (must be in range 0..6): -1
Return	false
Log	Invalid default runlevel (must be in range 0..6): 7
Return	false
Execute	.target.bash "/bin/ln -s -f /lib/systemd/system/runlevel3.target /etc/systemd/system/default.target" 0
Return	true
Execute	.target.bash "/bin/ln -s -f /lib/systemd/system/runlevel5.target /etc/systemd/system/default.target" 0
Return	true
Execute	.target.bash "/bin/ln -s -f /lib/systemd/system/runlevel5.target /etc/systemd/system/default.target" 1
Return	false
Dump	Systemd::DefaultRunlevel() tests
Read	.target.symlink "/etc/systemd/system/default.target" nil
Log	Cannot read symlink target of /etc/systemd/system/default.target
Return	nil
Read	.target.symlink "/etc/systemd/system/default.target" "/lib/systemd/system/runlevel3.target"
Return	3
Read	.target.symlink "/etc/systemd/system/default.target" "/lib/systemd/system/graphical.target"
Return	5
Read	.target.symlink "/etc/systemd/system/default.target" "/lib/systemd/system/unknown_default.target"
Return	nil
